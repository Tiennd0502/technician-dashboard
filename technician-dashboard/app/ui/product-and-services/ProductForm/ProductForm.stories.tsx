import type { Meta, StoryObj } from '@storybook/react';
import { useArgs } from '@storybook/preview-api';

import ProductForm from '.';

const meta: Meta<typeof ProductForm> = {
  title: 'Components/ProductForm',
  component: ProductForm,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof ProductForm>;

export const Default: Story = {
  args: {
    title: 'Add product',
    isOpen: true,
  },
  render: function Render(props) {
    const [{ isOpen }, updateArgs] = useArgs();

    const handleClose = () => updateArgs({ isOpen: !isOpen });

    return <ProductForm {...props} isOpen={isOpen} onClose={handleClose} />;
  },
};
