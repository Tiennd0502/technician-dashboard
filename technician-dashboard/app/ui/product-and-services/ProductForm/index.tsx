'use client';

import React, { memo, useCallback } from 'react';
import { Controller, useForm } from 'react-hook-form';
import {
  Button,
  Flex,
  FormControl,
  FormLabel,
  Radio,
  RadioGroup,
  VStack,
  useToast,
} from '@chakra-ui/react';
import { useRouter } from 'next/navigation';

// Types
import { Product, STATUS } from '@/lib/interfaces';

// Constants
import { PRODUCT_RULES } from '@/lib/constants';

// Hooks
import { useProduct } from '@/lib/hooks';

// Components
import { Input } from '@/ui/components';

interface ProductFromProps {
  product?: Product;
}

const ProductFrom = ({ product }: ProductFromProps) => {
  const router = useRouter();
  const toast = useToast();

  const {
    createProduct: { mutate: createProduct, isPending: isCreating },
    editProduct: { mutate: editProduct, isPending: isEditing },
  } = useProduct();

  const { control, handleSubmit } = useForm<Product>({
    reValidateMode: 'onSubmit',
    defaultValues: {
      ...product,
      status: product?.status || STATUS.Deactivated,
    },
  });

  const handleClose = useCallback(() => router.back(), [router]);

  const handleSubmitProduct = useCallback(
    (data: Product) => {
      const isUpdate = product?.id;
      const mutate = isUpdate ? editProduct : createProduct;

      mutate(data, {
        onSuccess: () => {
          toast({
            title: `Product ${isUpdate ? 'updated' : 'created'}.`,
            status: 'success',
            duration: 9000,
            isClosable: true,
          });
          router.refresh();
        },
        onError: () => {
          toast({
            title: `Failed to ${isUpdate ? 'update' : 'create'} the product.`,
            status: 'error',
            duration: 9000,
            isClosable: true,
          });
        },
      });
    },
    [product?.id, editProduct, createProduct, toast, router],
  );

  const isLoading = isCreating || isEditing;

  return (
    <VStack as='form' onSubmit={handleSubmit(handleSubmitProduct)} id='form' spacing='2' w='full'>
      <Controller
        control={control}
        name='name'
        rules={PRODUCT_RULES.name}
        render={({ field: { value, ...field }, fieldState: { error } }) => (
          <Input
            {...field}
            error={error?.message || ''}
            label='Product Name:'
            defaultValue={value}
            placeholder='Product Name'
          />
        )}
      />
      <Controller
        control={control}
        name='brand'
        rules={PRODUCT_RULES.service}
        render={({ field: { value, ...field }, fieldState: { error } }) => (
          <Input
            {...field}
            error={error?.message || ''}
            label='Brand Name:'
            defaultValue={value}
            placeholder='Brand Name'
          />
        )}
      />
      <Controller
        control={control}
        name='service'
        rules={PRODUCT_RULES.service}
        render={({ field: { value, ...field }, fieldState: { error } }) => (
          <Input
            {...field}
            error={error?.message || ''}
            label='Service:'
            defaultValue={value}
            placeholder='Service'
          />
        )}
      />
      <Controller
        control={control}
        name='status'
        render={({ field: { value, ...field } }) => (
          <FormControl>
            <FormLabel>Status:</FormLabel>
            <RadioGroup
              defaultValue={value?.toString() || STATUS.Deactivated.toString()}
              {...field}
            >
              <Flex gap='3'>
                <Radio {...field} value={STATUS.Activated.toString()}>
                  Activated
                </Radio>
                <Radio value={STATUS.Deactivated.toString()}>Deactivated</Radio>
              </Flex>
            </RadioGroup>
          </FormControl>
        )}
      />
      <Flex w='full'>
        <Button type='submit' form='form' ml='auto' mr='6' isLoading={isLoading}>
          Submit
        </Button>
        <Button variant='outline' isDisabled={isLoading} onClick={handleClose}>
          Cancel
        </Button>
      </Flex>
    </VStack>
  );
};

export default memo(ProductFrom);
