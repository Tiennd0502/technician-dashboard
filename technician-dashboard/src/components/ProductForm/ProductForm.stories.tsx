import type { Meta, StoryObj } from '@storybook/react';

import ProductForm from '.';

const meta: Meta<typeof ProductForm> = {
  title: 'Components/ProductForm',
  component: ProductForm,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof ProductForm>;

export const Default: Story = {
  args: {},
};
