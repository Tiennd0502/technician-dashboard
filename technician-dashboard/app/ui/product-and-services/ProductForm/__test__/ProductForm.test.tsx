import { render } from '@testing-library/react';
import { QueryClientProvider, QueryClient } from '@tanstack/react-query';

import ProductForm from '..';

const queryClient = new QueryClient();

const onClose = jest.fn();

const mockMutateAsync = jest.fn();

jest.mock('@/lib/hooks', () => ({
  ...jest.requireActual('@/lib/hooks'),
  useProduct: () => ({
    createProduct: {
      mutateAsync: mockMutateAsync,
      isLoading: false,
    },
    editProduct: {
      mutateAsync: mockMutateAsync,
      isLoading: false,
    },
  }),
}));

describe('ProductForm', () => {
  const setup = () =>
    render(
      <QueryClientProvider client={queryClient}>
        <ProductForm isOpen={true} onClose={onClose} />
      </QueryClientProvider>,
    );

  it('it matches the snapshot', () => {
    const dom = setup();
    expect(dom).toMatchSnapshot();
  });

  test('should simulate click button Submit', () => {
    const dom = setup();
    const btnClick = dom.getByRole('button', { name: 'Submit' });

    expect(btnClick).toBeTruthy();
  });
});
