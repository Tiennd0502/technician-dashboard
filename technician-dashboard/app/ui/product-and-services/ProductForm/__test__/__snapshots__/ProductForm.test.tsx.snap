// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm it matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
    />
    <div
      class="chakra-portal"
    >
      <div
        aria-hidden="true"
        class="chakra-modal__overlay css-13d1tkw"
        data-aria-hidden="true"
        style="opacity: 0;"
      />
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div
          class="chakra-modal__content-container css-1oxhx2p"
          tabindex="-1"
        >
          <section
            aria-describedby="chakra-modal--body-:r0:"
            aria-labelledby="chakra-modal--header-:r0:"
            aria-modal="true"
            class="chakra-modal__content css-17pp93c"
            id="chakra-modal-:r0:"
            role="dialog"
            style="opacity: 0; transform: scale(0.95) translateZ(0);"
            tabindex="-1"
          >
            <header
              class="chakra-modal__header css-72fd9l"
              id="chakra-modal--header-:r0:"
            >
              Add Product
            </header>
            <button
              aria-label="Close"
              class="chakra-modal__close-btn css-6sdaxy"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="chakra-modal__body css-0"
              id="chakra-modal--body-:r0:"
            >
              <form
                class="chakra-stack css-eb1ssy"
                id="form"
              >
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-gu124"
                    for="field-:r3:"
                    id="field-:r3:-label"
                  >
                    Product Name:
                  </label>
                  <input
                    class="chakra-input css-1cu97os"
                    id="field-:r3:"
                    name="name"
                    placeholder="Product Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-gu124"
                    for="field-:r4:"
                    id="field-:r4:-label"
                  >
                    Brand Name:
                  </label>
                  <input
                    class="chakra-input css-1cu97os"
                    id="field-:r4:"
                    name="brand"
                    placeholder="Brand Name"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-gu124"
                    for="field-:r5:"
                    id="field-:r5:-label"
                  >
                    Service:
                  </label>
                  <input
                    class="chakra-input css-1cu97os"
                    id="field-:r5:"
                    name="service"
                    placeholder="Service"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="chakra-form-control css-0"
                  role="group"
                >
                  <label
                    class="chakra-form__label css-uvho5g"
                    for="field-:r6:"
                    id="field-:r6:-label"
                  >
                    Status:
                  </label>
                  <div
                    class="chakra-radio-group css-0"
                    role="radiogroup"
                  >
                    <div
                      class="css-blvggq"
                    >
                      <label
                        class="chakra-radio css-1nqw95c"
                      >
                        <input
                          class="chakra-radio__input"
                          id="radio-:r8:"
                          name="status"
                          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                          type="radio"
                          value="1"
                        />
                        <span
                          aria-hidden="true"
                          class="chakra-radio__control css-ssalds"
                        />
                        <span
                          class="chakra-radio__label css-1y8kf23"
                        >
                          Activated
                        </span>
                      </label>
                      <label
                        class="chakra-radio css-1nqw95c"
                        data-checked=""
                      >
                        <input
                          checked=""
                          class="chakra-radio__input"
                          id="radio-:r9:"
                          name="status"
                          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
                          type="radio"
                          value="0"
                        />
                        <span
                          aria-hidden="true"
                          class="chakra-radio__control css-ssalds"
                          data-checked=""
                        />
                        <span
                          class="chakra-radio__label css-1y8kf23"
                          data-checked=""
                        >
                          Deactivated
                        </span>
                      </label>
                    </div>
                  </div>
                </div>
                <div
                  class="css-4v9c5a"
                >
                  <button
                    class="chakra-button css-oglbjv"
                    form="form"
                    type="submit"
                  >
                    Submit
                  </button>
                  <button
                    class="chakra-button css-4xx2wk"
                    type="button"
                  >
                    Cancel
                  </button>
                </div>
              </form>
            </div>
          </section>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
    data-aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
